core/src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java

```diff
@@ -142,10 +142,20 @@ public long valueForKey(long time) {
 
         @Override
         public long nextRoundingValue(long time) {
-            long timeLocal = time;
-            timeLocal = timeZone.convertUTCToLocal(time);
-            long nextInLocalTime = durationField.add(timeLocal, 1);
-            return timeZone.convertLocalToUTC(nextInLocalTime, false);
+            if (unitIsDayOrLonger()) {
+                time = timeZone.convertUTCToLocal(time);
+            }
+            long next = durationField.add(time, 1);
+            if (unitIsDayOrLonger()) {
+                return timeZone.convertLocalToUTC(next, false);
+            } else {
+                return next;
+            }
+        }
+
+        private boolean unitIsDayOrLonger() {
+            return (unit == DateTimeUnit.HOUR_OF_DAY || unit == DateTimeUnit.MINUTES_OF_HOUR
+                    || unit == DateTimeUnit.SECOND_OF_MINUTE) == false;
```

Comment:
I'm not sure we should to do any arithmetic only to check what kind of enum we have. There are only eight options, checking if any of the three are set seems clearer to me, wdyt?